hibernate-core/src/main/java/org/hibernate/dialect/FirebirdDialect.java

```diff
@@ -219,24 +222,87 @@ public SqlAstTranslatorFactory getSqlAstTranslatorFactory() {
 	 */
 	@Override
 	public String castPattern(CastType from, CastType to) {
-		if ( to==BOOLEAN
-				&& (from==LONG || from==INTEGER)) {
-			return "(0<>?1)";
-		}
-		if ( getVersion() < 300 ) {
-			if ( to==BOOLEAN && from==STRING ) {
-//				return "iif(lower(?1) similar to 't|f|true|false', lower(?1) like 't%', null)";
-				return "decode(lower(?1),'t',1,'f',0,'true',1,'false',0)";
-			}
-			if ( to==STRING && from==BOOLEAN ) {
-				return "trim(decode(?1,0,'false','true'))";
-			}
-		}
-		else {
-			if ( from==BOOLEAN
-					&& (to==LONG || to==INTEGER)) {
-				return "decode(?1,false,0,true,1)";
-			}
+		switch ( to ) {
+			case FLOAT:
+				return "cast(double(?1) as real)";
+			case DOUBLE:
+				return "double(?1)";
+			case INTEGER:
+			case LONG:
+				switch ( from ) {
+					case BOOLEAN:
+						return "decode(?1,false,0,true,1,null)";
+					case YN_BOOLEAN:
+						return "decode(?1,'Y',1,'N',0,null)";
+					case TF_BOOLEAN:
+						return "decode(?1,'T',1,'F',0,null)";
+				}
+				break;
+			case BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "decode(?1,'T',true,'F',false,'Y',true,'N',false,null)";
+					case YN_BOOLEAN:
+						return "decode(?1,'Y',true,'N',false,null)";
+					case TF_BOOLEAN:
+						return "decode(?1,'T',true,'F',false,null)";
+					case INTEGER:
+					case LONG:
+					case INTEGER_BOOLEAN:
+						return "decode(?1,abs(sign(?1),1,true,0,false,null)";
+				}
+				break;
+			case INTEGER_BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "decode(?1,'T',1,'F',0,'Y',1,'N',0,null)";
+					case YN_BOOLEAN:
+						return "decode(?1,'Y',1,'N',0,null)";
+					case TF_BOOLEAN:
+						return "decode(?1,'T',1,'F',0,null)";
+					case INTEGER:
+					case LONG:
+						return "abs(sign(?1))";
+				}
+				break;
+			case YN_BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "decode(?1,'T','Y','F','N','Y','Y','N','N',null)";
+					case INTEGER_BOOLEAN:
+						return "decode(?1,1,'Y',0,'N',null)";
+					case TF_BOOLEAN:
+						return "decode(?1,'T','Y','F','N',null)";
+					case INTEGER:
+					case LONG:
+						return "decode(abs(sign(?1),1,'Y',0,'N',null)";
+				}
+				break;
+			case TF_BOOLEAN:
+				switch ( from ) {
+					case STRING:
+						return "decode(?1,'T','Y','F','N','Y','Y','N','N',null)";
+					case INTEGER_BOOLEAN:
+						return "decode(?1,1,'T',0,'F',null)";
+					case YN_BOOLEAN:
+						return "decode(?1,'Y','T','N','F',null)";
+					case INTEGER:
+					case LONG:
+						return "decode(abs(sign(?1),1,'T',0,'F',null)";
+				}
+				break;
+			case STRING:
+				switch ( from ) {
+					case INTEGER_BOOLEAN:
+						return "decode(?1,null,null,0,'false','true')";
+					case TF_BOOLEAN:
+						return "decode(?1,'T','true','F','false',null)";
+					case YN_BOOLEAN:
+						return "decode(?1,'Y','true','N','false',null)";
+					case BOOLEAN:
+						return "decode(?1,true,'true',false,'false',null)";
+				}
+				break;
```

Comment:
I suppose this is just a copy/paste job of what's in `OracleDialect`. If so, it should be moved to a static method somewhere, to eliminate the duplication, and also to reuse it in HSQLDB.